src/core/sm/launcher/launcher.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (C) 2025 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <core/common/tools/logger.hpp>
+
+#include "launcher.hpp"
+
+namespace aos::sm::launcher {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Error Launcher::Init(const Array<RuntimeItf*> runtimes, StorageItf& storage, InstanceStatusSenderItf& statusSender,
+    InstanceStatusProviderItf& statusProvider)
+{
+    LOG_INF() << "Initializing launcher";
+
+    for (auto* runtime : runtimes) {
+        if (auto err = mRuntimes.Set(runtime, ""); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    mStorage        = &storage;
+    mStatusSender   = &statusSender;
+    mStatusProvider = &statusProvider;
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::Start()
+{
+    LockGuard lock {mMutex};
+
+    LOG_INF() << "Start launcher";
+
+    RuntimeInfo runtimeInfo;
+
+    for (auto& it : mRuntimes) {
+        if (auto err = it.mFirst->GetInfo(runtimeInfo); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+
+        if (auto err = it.mSecond.Assign(runtimeInfo.mRuntimeID); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+
+        LOG_INF() << "Registered runtime" << Log::Field("runtimeID", runtimeInfo.mRuntimeID)
+                  << Log::Field("type", runtimeInfo.mRuntimeType);
+    }
+
+    if (auto err = mStatusProvider->SubscribeListener(*this); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (auto err = StartStoredInstances(); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::Stop()
+{
+    LockGuard lock {mMutex};
+
+    LOG_INF() << "Stop launcher";
+
+    mStatusProvider->UnsubscribeListener(*this);
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::UpdateInstances(const Array<InstanceIdent>& stopInstances, const Array<InstanceInfo>& startInstances)
+{
+    LockGuard lock {mMutex};
+
+    return UpdateInstancesImpl(stopInstances, startInstances);
```

Comment:
It means we will block here till all instances are processed it could take some timer. This call comes from grpc client. I wonder if it will not block other grpc communications.